pkg/fqdn/dnspoller.go

```diff
@@ -0,0 +1,373 @@
+			continue perRule
+		}
+
+		uuidLabel := generateUUIDLabel()
+		newDNSNames, alreadyExistsDNSNames := poller.addRule(uuidLabel.Value, sourceRuleCopy)
+
+		// only add a custom label (below) if we processed any ToFQDN rules.
+		if (len(newDNSNames) + len(alreadyExistsDNSNames)) > 0 {
+			// add a unique ID that we can use later to replace this rule.
+			sourceRuleCopy.Labels = append(sourceRuleCopy.Labels, uuidLabel)
+
+			log.WithFields(logrus.Fields{
+				"newDNSNames":           newDNSNames,
+				"alreadyExistsDNSNames": alreadyExistsDNSNames,
+				"numRules":              len(sourceRules),
+			}).Debug("Added FQDN to poll list")
+		}
+	}
+
+	return markedRules, nil
+}
+
+// StopPollForDNSName runs the bookkeeping to remove each api.Rule from
+// corresponding dnsName entries in sourceRules and IPs.  When no more rules
+// rely on a specific dnsName, we remove it from the maps and stop polling for
+// it. It expects rules to still be labelled with a ToFQDN-UUID if
+// StartPollForDNSName had added the label on insertion.
+// Note: rule deletion in policy.Repository is by label, where the rules must
+// have at least the labels in the delete.  This means our ToFQDN-UUID label,
+// and later ToCIDRSet additions will also be deleted correctly, and no action
+// is needed here.
+func (poller *DNSPoller) StopPollForDNSName(sourceRules []*api.Rule) {
+	poller.Lock()
+	defer poller.Unlock()
+
+	for _, sourceRule := range sourceRules {
+		// skip unmarked rules, nothing to do
+		if !sourceRule.Labels.Has(uuidLabelSearchKey) {
+			continue
+		}
+
+		uuid := getUUIDFromRuleLabels(sourceRule)
+		noLongerPolledDNSNames := poller.removeRule(uuid, sourceRule)
+		log.WithFields(logrus.Fields{
+			"noLongerPolled": noLongerPolledDNSNames,
+		}).Debug("Removed FQDN from poll list")
+	}
+}
+
+// LookupUpdateDNS runs a DNS lookup for each stored DNS name, storing updates,
+// and then emits regenerated policy rules.
+// The general steps are:
+// 1- take a snapshot of DNS names to lookup from poller, into dnsNamesToPoll
+// 2- Do a DNS lookup for each DNS name (map key) in poller via LookupDNSNames
+// 3- Update IPs for each dnsName in poller. If the IPs have changed for the
+// name, store which rules must be updated in rulesToUpdate. This is a set and
+// is deduped
+// 4- For each rule in rulesToUpdate, generate a new policy rule with IPs
+// 5- If we have any rules to update, emit them with AddGeneratedRules
+func (poller *DNSPoller) LookupUpdateDNS() error {
+	// Collect the DNS names that need lookups. This avoids locking
+	// poller during lookups.
+	dnsNamesToPoll := poller.GetDNSNames()
+
+	// lookup the DNS names. Names with failures will not be updated (and we
+	// will use the most recent data below)
+	updatedDNSIPs, errorDNSNames := poller.config.LookupDNSNames(dnsNamesToPoll)
+	for dnsName, err := range errorDNSNames {
+		log.WithError(err).WithField("matchName", dnsName).Warn("Cannot resolve FQDN")
```

Comment:
Might be useful to also include in the message the implication.

I believe the implication is something like "Traffic egressing to this destination may be dropped".